lms/views/decorators/h_api.py

```diff
@@ -0,0 +1,126 @@
+"""View decorators for working with the h API."""
+
+import functools
+
+from pyramid.httpexceptions import HTTPBadRequest
+
+from lms.services import HAPINotFoundError
+
+
+__all__ = ["upsert_h_user", "upsert_course_group", "add_user_to_group"]
+
+
+def upsert_h_user(wrapped):
+    """
+    Create or update the Hypothesis user for the request's LTI user.
+
+    Update the h user's information from LTI data. If the user doesn't exist
+    yet, call the h API to create one.
+
+    Assumes that it's only used on LTI launch views.
+    """
+
+    @functools.wraps(wrapped)
+    def wrapper(context, request):
+        if not context.provisioning_enabled:
+            return wrapped(context, request)
+
+        hapi_svc = request.find_service(name="hapi")
+
+        user_data = {
+            "username": context.h_username,
+            "display_name": context.h_display_name,
+            "authority": request.registry.settings["h_authority"],
+            "identities": [
+                {
+                    "provider": context.h_provider,
+                    "provider_unique_id": context.h_provider_unique_id,
+                }
+            ],
+        }
+
+        try:
+            hapi_svc.patch(
+                f"users/{context.h_username}", {"display_name": context.h_display_name}
+            )
+        except HAPINotFoundError:
+            # Call the h API to create the user in h if it doesn't exist already.
+            hapi_svc.post("users", user_data)
+
+        return wrapped(context, request)
+
+    return wrapper
+
+
+def upsert_course_group(wrapped):
+    """
+    Create or update the Hypothesis group for the request's LTI course.
+
+    Call the h API to create a group for the LTI course, if one doesn't exist
+    already.
+
+    Groups can only be created if the LTI user is allowed to create Hypothesis
+    groups (for example instructors are allowed to create groups). If the group
+    for the course hasn't been created yet, and if the user isn't allowed to
+    create groups (e.g. if they're just a student) then show an error page
+    instead of continuing with the LTI launch.
+
+    Assumes that it's only used on LTI launch views.
+    """
+
+    @functools.wraps(wrapped)
+    def wrapper(context, request):
+        if not context.provisioning_enabled:
+            return wrapped(context, request)
+
+        hapi_svc = request.find_service(name="hapi")
+
+        is_instructor = any(
```

Comment:
Does this "question" get asked anywhere else in the codebase? Is it worth util-izing at some point? Likely premature optimization at this point...

Reply:
Yeah this needs moving somewhere, possible into the `LTILaunch` context or into `LTIUser`